pkg/election/election.go

```diff
@@ -295,12 +358,116 @@ func (e *ZooKeeperElection) BecomeLeader(ctx context.Context) error {
 		}
 	}
 
+	e.claimLeaderRole(ctx)
+
 	return nil
 }
 
+func convertMsEpochToTime(ms int64) time.Time {
+	return time.Unix(ms/1000, (ms%1000)*1_000_000)
+}
+
+func (e *ZooKeeperElection) hasActiveLeader() bool {
+	e.log.Info("checking if there is an active leader")
+
+	data, stat, err := e.zk.Get(e.basePath)
+	if err != nil {
+		e.log.Errorf("could not get leader node %s, assuming another leader is active: %v", e.basePath, err)
+
+		// if we can't reach zk, assume there's already a
+		// leader, as we don't want to risk a split brain
+		// election
+		return true
+	}
+
+	mtime := convertMsEpochToTime(stat.Mtime)
+	since := time.Since(mtime)
+	hasActiveLeader := since < e.sessionTimeout && len(data) > 0
+
+	e.log.Debugf("active leader info: hasActiveLeader=%v descr=%s since=%s threshold=%s",
+		hasActiveLeader, string(data), since.String(), e.sessionTimeout.String())
+
+	if hasActiveLeader {
+		e.log.Infof("active leader '%s' was seen %s ago", string(data), since.String())
+	}
+
+	return hasActiveLeader
+}
+
+func (e *ZooKeeperElection) clearLeaderRole() {
+	e.lock("clearLeaderRole")
+	defer e.unlock("clearLeaderRole")
+
+	_, stat, err := e.zk.Get(e.basePath)
+	if err != nil {
+		e.log.Errorf("could not get leader node version: %v", err)
+	}
+
+	err = e.zk.Set(e.basePath, []byte{}, stat.Version)
+	if err != nil {
+		e.log.Errorf("could not clear leader role: %v", err)
+	}
+}
+
+func (e *ZooKeeperElection) publishLeaderRole() {
+	e.lock("publishLeaderRole")
+	defer e.unlock("publishLeaderRole")
+
+	if e.proposalNodePath == "" {
+		e.log.Debug("in the process of resigning, not publishing")
+
+		return
+	}
+
+	e.log.Debug("publishing leader state")
+
+	_, stat, err := e.zk.Get(e.basePath)
+	if err != nil {
+		e.log.Errorf("could not get leader node version: %v", err)
+
+		return
+	}
+
+	err = e.zk.Set(e.basePath, e.serializeLeaderInfoUnlocked(), stat.Version)
```

Comment:
is passing `stat.Version` a condition on the version? I am wondering why it is needed or if we could remove the version condition and the previous `zk.Get` call.

Reply:
Good catch, it was used early in the implementation to make sure that no 2 sessions thought they're leader, but it was not robust enough to be used on its own.